haskell/toolchain.bzl

```diff
@@ -21,10 +39,75 @@ def _haskell_toolchain_impl(ctx):
     arguments = [arguments],
   )
 
+  # NOTE The only way to let various executables know where other
+  # executables are located is often only via the PATH environment variable.
+  # For example, ghc uses gcc which is found on PATH. This forces us provide
+  # correct PATH value (with e.g. gcc on it) when we call executables such
+  # as ghc. However, this hurts hermeticity because if we construct PATH
+  # naively, i.e. through concatenation of directory names of all
+  # executables of interest, we also make other neighboring executables
+  # visible, and they indeed can influence the building process, which is
+  # undesirable.
+  #
+  # To prevent this, we create a special directory populated with symbolic
+  # links to all executables we want to make visible, and only to them. Then
+  # we make all rules depend on some of these symlinks, and provide PATH
+  # (always the same) that points only to this prepared directory with
+  # symlinks. This helps hermeticity and forces us to be explicit about all
+  # programs we need/use for building. Setting PATH and listing as inputs
+  # are both necessary for a tool to be available with this approach.
+
+  targets = [
+    # CPP host fragments.
+    ctx.host_fragments.cpp.ar_executable,
+    ctx.host_fragments.cpp.compiler_executable,
+    ctx.host_fragments.cpp.ld_executable,
+    ctx.host_fragments.cpp.nm_executable,
+    ctx.host_fragments.cpp.preprocessor_executable,
+    ctx.host_fragments.cpp.strip_executable,
+    # Binutils we grab from the attributes. Hopefully we'll be able to use
+    # fragments for this too once
+    #
+    # https://github.com/bazelbuild/bazel/issues/4681
+    #
+    # is merged.
+    ctx.attr.ln_location,
+    ctx.attr.grep_location,
+  ] + ghc_binaries # Previously collected GHC binaries.
+
+  # Create symlinks to binaries.
+  symlinks = set.empty()
+  for target in targets:
+    symlink = ctx.actions.declare_file(
+      paths.join("visible-binaries", paths.basename(target))
+    )
+    ctx.actions.run(
+      inputs = ctx.files.tools,
+      outputs = [symlink],
+      executable = ctx.file._make_bin_symlink,
+      arguments = [
+        ctx.attr.mkdir_location,
+        ctx.attr.realpath_location,
+        ctx.attr.ln_location,
+        target,
+        symlink.dirname,
+        symlink.path,
+      ],
+    )
+    set.mutable_insert(symlinks, symlink)
+
+  tools_struct_args = {tool.basename.replace("-", "_"): tool
```

Comment:
why is this replace needed?

Reply:
Because you have binaries named like `ghc-pkg` and `ghc-pkg` is not a valid attribute name, so it must be `ghc_pkg`.